### DISCUSSION TITLE
Error in readGDX(gdx, "f10_land", "f_land", format = "first_found") :   could not find function "readGDX"

### DISCUSSION THREAD

**Discussion by CEEEJAYLEE** [magpiemodel/magpie#709] [Help] [answered]
Dear All,

I have been trying to run the default model but encountered this error every time I ran the model, with some subsequent errors in output. Any idea what is going on? Below is the code generated:

NPI/NDC recalculation successful!
2024-08-05 11:35:41.168429: unlocked model.
Saving timePrepareStart and timePrepareEnd to runstatistics.rda

Starting MAgPIE...
[1] 0

MAgPIE run finished!
Model finished with modelstat 2:Local optimal solution achieved.
**Cannot access run statistics repository /p/projects/rd3mod/models/statistics/magpie Run statistics are not submitted.**
Generating lockfile... done.
Lockfile written to D:\magpie\output\default_2024-08-05_11.27.31/2024-08-05T115926__output_check--extra-disaggregation--rds_report__renv.lock

# output_check.R -> D:/magpie/output/default_2024-08-05_11.27.31
Output looks good, no problems found!

# extra/disaggregation.R -> D:/magpie/output/default_2024-08-05_11.27.31

Attaching package: 'dplyr'

The following object is masked from 'package:magclass':

    where

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union

**Error in readGDX(gdx, "f10_land", "f_land", format = "first_found") :
  could not find function "readGDX"**

# rds_report.R -> D:/magpie/output/default_2024-08-05_11.27.31
Start getReport(gdx)...
   reportPopulation(gdx)                                     success 0.58s
   reportWorkingAgePopulation(gdx)                           success 0.31s
   reportIncome(gdx)                                         success 0.08s
   reportPriceGHG(gdx)                                       success 0.78s
   reportFoodExpenditure(gdx)                                success 2.53s
   reportKcal(gdx,detail=detail)                             success 0.61s
   reportProtein(gdx,detail=detail)                          success 0.65s
   reportIntakeDetailed(gdx,detail=detail)                   success 0.84s
   reportAnthropometrics(gdx)                                success 6.91s
   reportLivestockShare(gdx)                                 success 0.30s
   reportLivestockDemStructure(gdx)                          success 0.38s
   reportVegfruitShare(gdx)                                  success 0.28s
   **reportPriceShock(gdx)                                     no return value 0.02s**
   reportPriceElasticities(gdx)                              success 3.34s
   reportDemand(gdx,detail=detail)                           success 2.83s
   reportDemandBioenergy(gdx,detail=detail)                  success 1.89s
   reportFeed(gdx,detail=detail)                             success 1.88s
   reportProduction(gdx,detail=detail)                       success 0.27s
   reportProductionBioenergy(gdx,detail=detail)              success 0.31s
   reportTrade(gdx,detail=detail)                            success 4.55s
   reportLandFootprint(gdx)                                  success 1.70s
   reportLandUse(gdx)                                        success 0.56s
   reportLandUseChange(gdx)                                  success 0.54s
   reportNetForestChange(gdx)                                success 0.17s
   reportPeatland(gdx)                                       success 0.14s
   reportProtectedArea(gdx)                                  success 0.06s
   reportCroparea(gdx,detail=detail)                         success 0.78s
   reportNitrogenBudgetCropland(gdx, dir = dir)              success 1.44s
   reportNitrogenBudgetPasture(gdx, dir = dir)               success 0.23s
   reportNitrogenEfficiencies(gdx)                           success 1.83s
   reportManure(gdx)                                         success 0.17s
   **reportNitrogenPollution(gdx, dir = dir)                   ERROR 4.20s**
   reportYields(gdx,detail=detail)                           success 1.80s
   reportYieldsCropCalib(gdx,detail=detail)                  success 1.71s
   reportYieldsCropRaw(gdx,detail=detail)                    success 2.06s
   reportFeedConversion(gdx)                                 success 0.77s
   reportTau(gdx)                                            success 0.34s
   reportTc(gdx)                                             success 0.55s
   reportAgriResearchIntensity(gdx)                          success 0.13s
   reportEmissions(gdx)                                      success 146.78s
   reportEmissionsBeforeTechnicalMitigation(gdx)             success 0.42s
   reportCosts(gdx)                                          success 2.67s
   reportCostsPresolve(gdx)                                  no return value 0.00s
   reportCostTransport(gdx)                                  success 0.31s
   reportCostsFertilizer(gdx)                                success 0.28s
   reportCostOverall(gdx)                                    success 0.11s
   reportCostCapitalStocks(gdx)                              no return value 0.00s
   reportCostCapitalInvestment(gdx)                          no return value 0.02s
   reportCostInputsCrop(gdx)                                 success 0.11s
   reportCostsInputFactors(gdx)                              success 0.55s
   reportCostsAccounting(gdx)                                success 5.73s
   reportCostsProductionCrops(gdx,type = 'investment')       success 5.56s
   reportCostsWithoutIncentives(gdx)                         success 5.93s
   reportAgGDP(gdx)                                          success 1.19s
   reportConsumVal(gdx)                                      success 0.54s
   reportPriceFoodIndex(gdx,baseyear = 'y2010')              success 3.41s
   reportPriceFoodIndex(gdx,baseyear = 'y2020')              success 3.48s
   reportProducerPriceIndex(gdx)                             success 0.96s
   reportExpenditureFoodIndex(gdx)                           success 4.14s
   reportPriceAgriculture(gdx)                               success 0.14s
   reportPriceBioenergy(gdx)                                 success 0.15s
   reportPriceLand(gdx)                                      success 0.20s
   reportPriceWater(gdx)                                     success 0.17s
   reportValueTrade(gdx)                                     success 3.08s
   reportProcessing(gdx, indicator='primary_to_process')     success 0.25s
   reportProcessing(gdx, indicator='secondary_from_primary') success 0.67s
   reportAEI(gdx)                                            success 0.09s
   reportWaterUsage(gdx)                                     ERROR 0.03s
   reportWaterAvailability(gdx)                              success 0.05s
   reportAAI(gdx)                                            success 0.16s
   reportSOM(gdx)                                            success 1.21s
   reportGrowingStock(gdx, indicator='relative')             success 1.95s
   reportGrowingStock(gdx, indicator='absolute')             success 1.44s
   reportSDG1(gdx)                                           success 0.06s
   reportSDG2(gdx)                                           success 8.62s
   reportSDG3(gdx)                                           success 4.19s
   reportSDG6(gdx)                                           ERROR 1.39s
   reportSDG12(gdx)                                          success 2.83s
   reportSDG15(gdx)                                          success 2.31s
   reportForestYield(gdx)                                    success 0.70s
   reportharvested_area_timber(gdx)                          success 0.64s
   reportPlantationEstablishment(gdx)                        success 0.25s
   reportRotationLength(gdx)                                 success 0.11s
   reportTimber(gdx)                                         success 0.47s
   **reportBII(gdx, dir = dir)                                 ERROR 0.36s**
   reportCropDiversity(gdx)                                  success 0.78s
   reportPriceWoodyBiomass(gdx)                              success 0.28s
   reportCarbonstock(gdx)                                    success 0.21s
   reportGrasslandManagement(gdx)                            Disabled (no managed pastures) 0.05s
   reportGrassStats(gdx)                                     Disabled (no managed pastures) 0.00s
   reportGrasslandYields(gdx)                                Disabled (No separate grassland yields) 0.04s
   reportLSUGrasslands(gdx)                                  Disabled (No LSU management) 0.02s
   reportAgEmployment(gdx, type = 'absolute', detail = TRUE) success 0.25s
   reportAgEmployment(gdx, type = 'share', detail = TRUE)    success 0.58s
   reportHourlyLaborCosts(gdx)                               success 0.45s
   reportRelativeHourlyLaborCosts(gdx)                       success 0.20s
   reportTotalHoursWorked(gdx)                               success 0.33s
   reportOutputPerWorker(gdx)                                success 0.58s
   reportValueMaterialDemand(gdx)                            success 0.37s
   reportFactorCostShares(gdx, type = 'requirements')        success 0.52s
   reportFactorCostShares(gdx, type = 'optimization')        success 0.53s
   reportFactorCostShares(gdx, type = 'accounting')          success 0.74s
   reportWageDevelopment(gdx, baseYear = 2000)               success 0.60s
   reportWageDevelopment(gdx, baseYear = 2010)               success 0.61s
   reportWageDevelopment(gdx, baseYear = 2020)               success 0.67s
   **reportWaterIndicators(gdx)                                ERROR 0.04s**
   reportBioplasticDemand(gdx)                               success 0.19s
   reportCostsMACCS(gdx)                                     success 0.11s
   reportLaborCostsEmpl(gdx)                                 success 0.39s
   reportLaborProductivity(gdx)                              no return value 0.00s
   reportRuralDemandShares(gdx, type = 'tradOnly')           no return value 0.00s
   reportCostsWholesale(gdx)                                 no return value 0.00s
   reportFit(gdx,type='R2',level='grid')                     ERROR 0.11s
   reportFit(gdx,type='R2',level='cell')                     success 19.68s
Total runtime:  332.03s

Warning message:
In runsubmit(output = output, alloutputdirs = outputdir, submit = submit,  :
  Script extra/disaggregation.R was stopped by an error and not executed properly!
Saving timeGAMSStart, timeGAMSEnd, timeOutputStart and timeOutputStart to runstatistics.rda
Warning message:
In runsubmit(output = output, alloutputdirs = outputdir,  ... :
  Script extra/disaggregation.R was stopped by an error and not executed properly!
2024-08-05 12:05:04.987672: unlocked model.
Warning messages:
1: In untar2(tarfile, files, list, exdir, restore_times) :
  using pax extended headers
2: In madrat::toolAggregate(x = magpie_policy, rel = rel, weight = weight) :
  Weight sum is 0, so cannot normalize and will return 0 for some aggregation targets. This changes the total sum of the magpie object! If this is really intended set zeroWeight = "allow".
3: In madrat::toolAggregate(x = magpie_policy, rel = rel, weight = weight) :
  Weight sum is 0, so cannot normalize and will return 0 for some aggregation targets. This changes the total sum of the magpie object! If this is really intended set zeroWeight = "allow".

**Comment by tscheypidi** [comment, marked as answer]
Hi, 

thanks for the error report. We recently replaced the gdx package dependency in magpie4 with gdx2. With this update the `readGDX` function is not automatically exported anymore by the `magpie4` package. It seems that the disaggregation script was using `readGDX` from these packages without explicitly loading any of these.

I created now a PR #710 adressing this issue. You can either wait for having this PR merged or just manually add `library(gdx2)` to the disaggregation output script as done [here](https://github.com/magpiemodel/magpie/pull/710/files#diff-8232dd52fa9ef237143de2adae9675f3da40233093f4b0415b47bd51951ac14aR19)

**Reply by CEEEJAYLEE** [reply]
Thanks very much for your help! Will try to figure that out.
